test: introduce tag functions to handle long strings in code

vsemozhetbyt opened the pull request:
##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
test

Multiline template literals in ES6 make long strings handling easier in code. However, there are many annoying nits that prevent this option from using in many cases: code block indentation adds unwanted spaces to these strings, leading and trailing backticks break alignment or add unwanted leading and trailing line breaks.

This PR proposes 4 helper functions to mitigate these difficulties:

1. `common.tagGlue()` — joins a multiline template literal seamlessly, i.e. removes all literal line breaks and leading spaces (handy to code big text blobs not intended for human reading).
2. `common.tagUnwrap()` — joins a multiline template literal with a space, i.e. replaces all literal line breaks and leading spaces with a space, then trims one leading and one trailing space (handy to code long readable text lines).
3. `common.tagLFy()` — preserves line breaks as `\n`, but removes two framing line breaks and only auxiliary indents (handy to code multiline readable text, JavaScript metacode, HTML markup etc).
4. `common.tagCRLFy()` — same as `tagLFy()`, but replaces all literal `\n` by `\r\n` (handy to code raw HTTP requests/responses).

You can see examples in the [proposed doc fragments](https://github.com/vsemozhetbyt/node/blob/306f2287ca5f40e85430ea988b3f74f1e55160d7/test/common/README.md#spawnsyncpwdoptions).

This PR also demonstrates the applying these functions to many test fragments.

All these functions leave interpolated variables untouched, only literal parts are processed. In most cases, these functions get multiline strings with leading and trailing line breaks: this allows to divide code noise (backticks etc) and pure data. These leading and trailing line breaks are stripped by the tag functions. This should be taken into account: if the result needs a line break in the very end, an additional line break should be added (beware, for example, HTTP raw code — see samples in the second commit).

I understand that this is a big PR and mostly based on personal taste. So see it as a strawman and feel free to reject without thorough explanations.

Though the diff seems huge, the changes are mostly trivial, concern multiline blocks formatting and are easy to skim.

vsemozhetbyt commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8061/

joyeecheung commented on test/common/index.js:
```diff
@@ -746,3 +749,47 @@ exports.getTTYfd = function getTTYfd() {
+  return expressions.reduce(
+    (acc, expr, i) => `${acc}${expr}${strings[i + 1].replace(breaks, ' ')}`,
+    strings[0].replace(breaks, ' ')
+  ).replace(/^ | $/g, '');
```
`.trim()`?

vsemozhetbyt replied:
On this stage, the replacing concerns the whole string, including interpolated variables, so I prefer to be extra careful here. i.e. to remove only two auxiliary spaces (which were framing line breaks before the reducing).

joyeecheung commented on test/common/index.js:
```diff
@@ -746,3 +749,47 @@ exports.getTTYfd = function getTTYfd() {
+
+// Removes two framing line breaks and only auxiliary indents
+exports.tagLF = function tagLF(strings, ...expressions) {
+  const [, firstIndent = ''] = strings[0].match(/\n+( *)/) || [];
```
If I am reading correctly `strings[0].match(/\n+( +)/)` should work as well?

vsemozhetbyt replied:
In case there are no indents — yes if I get it right, but maybe with `*` this returns earlier, without using the fallbacks `[]` and `''`

refack replied:
@joyeecheung would fail for (badly formated) literals like:
```js
const tagged = tagLF`
<!doctype html>
<html>
  <head>
    <meta charset="UTF-8">
    <title></title>
  </head>
  <body>
    ${process.versions}
  </body>
</html>
`;
```
would give `firstIndent === undefined`

refack commented on test/common/README.md:
```diff
@@ -261,6 +261,124 @@ Platform normalizes the `pwd` command.
+* return [&lt;String>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type)
+
+Tag function for tagged template literals:
+same as `tagLF()`, but replaces all literal `\n` by `\r\n`.
```
+ Useful for HTTP compliance

refack commented on test/common/README.md:
````diff
@@ -261,6 +261,124 @@ Platform normalizes the `pwd` command.
+assert.strictEqual(unTagged, tagged);
+```
+
+### tagCRLF()
````
[suggestion] maybe even rename `tagHTTP`

vsemozhetbyt replied:
There were other cases in the tests. And potentially can be others (Windows stuff etc).

refack commented on test/common/index.js:
```diff
@@ -211,9 +211,11 @@ Object.defineProperty(exports, 'localhostIPv4', {
-                      'as LOCALHOST or expect some tests to fail.');
+        console.error(exports.tagUnwrap`
+          Looks like we're in a FreeBSD Jail.
+          Please provide your default interface address
```
[suggestion] for better readability:
```js
        console.error(exports.tagUnwrap`
          Looks like we're in a FreeBSD Jail.
          Please provide your default interface address as LOCALHOST 
          or expect some tests to fail.
        `);
```

refack commented on test/common/index.js:
```diff
@@ -746,3 +749,47 @@ exports.getTTYfd = function getTTYfd() {
+
+// Removes all literal line breaks and leading spaces
+exports.tagGlue = function tagGlue(strings, ...expressions) {
+  const breaks = /\n */g;
```
move all consts out of the functions. no need to redefine on each invocation.
could be:
```js
const TAG_CONSTANTS = {
  tagGlueBreaks: /\n */g,
  tagUnwrapBreaks: /(\n *)+/g,
  tagUnwarpTrimmer: /^ | $/g,
...
```

vsemozhetbyt replied:
I would prefer to keep them self-sufficient while they are not used in libs critical for performance.

refack replied:
I won't block on this, but you might find out you can reuse some, or other coders might reuse, also gives them better names.

vsemozhetbyt replied:
I also remember that `common` module is considered too bloated, so I am afraid to pollute it with more global variables.

refack commented on test/common/index.js:
```diff
@@ -746,3 +749,47 @@ exports.getTTYfd = function getTTYfd() {
+// Removes two framing line breaks and only auxiliary indents
+exports.tagLF = function tagLF(strings, ...expressions) {
+  const [, firstIndent = ''] = strings[0].match(/\n+( *)/) || [];
+  const indent = new RegExp(`\n {0,${firstIndent.length}}`, 'g');
```
IMHO go stricter, i.e. ` new RegExp(`\n {${firstIndent.length}}`, 'g');`
example:
```js
const literlaHTML = tagLF`
  <!doctype html>
  <html>
    <head>
      <meta charset="UTF-8">
      <script>
var x = [1,2,3].map(functor, seed)
  .filter(predicate)
  .reduce(functor, '');
      <script>
      <title></title>
    </head>
    <body>
      ${process.versions}
    </body>
  </html>
`;
```

vsemozhetbyt replied:
I would prefer to ignore such misindentation cases in favor of these cases:
```js
{
  const tagged = tagLF`
    <!doctype html>
    <html>
      ...
    </html>
  `;
}
```
NB: indentation before the last backtick is less than the first indentation. With your variant, it will be preserved with the trailing line break.

vsemozhetbyt replied:
BTW, these misindented fragments can be wrapped in `${''}` and they will be untouched.

vsemozhetbyt commented on test/parallel/test-net-connect-options-allowhalfopen.js:
```diff
@@ -60,32 +62,38 @@ const forAllClients = (cb) => common.mustCall(cb, CLIENT_VARIANTS);
-                        `${clientReceivedFIN} FIN received by client, ` +
-                        `${clientSentFIN} FIN sent by client, ` +
-                        `${serverConnections} FIN sent by server`);
+          console.error(tagUnwrap`
```
OK.

vsemozhetbyt commented on test/parallel/test-promises-unhandled-rejections.js:
```diff
@@ -1,5 +1,5 @@
 'use strict';
```
I have only reformated the first string arguments in each function, nothing more. GitHub diff just goes astray here, sorry.

refack replied:
I'll take a look locally.

vsemozhetbyt commented on test/parallel/test-repl-persistent-history.js:
```diff
@@ -53,20 +55,38 @@ const ENTER = { name: 'enter' };
+
+`;
+const sameHistoryFilePaths = tagLF`
+
```
I will fix, thank you.

vsemozhetbyt replied:
I have recalled why: see the long line without a break in the previous variant, this is the only way to preserve it.

refack replied:
[suggestion] I know you don't like to change too much, but I would go
```js
const keepLF = '\n';
const fileName = path.join(common.tmpDir, '.node_repl_history');
const sameHistoryFilePaths = wrap`
  ${keepLF}
  The old repl history file has the same name and location as the new one i.e.,
  ${fileName} and is empty.
  Using it as is.
  ${keepLF}
`;
```
maybe even add `export.keepLF  = '\n'` to `common`

vsemozhetbyt replied:
I am afraid this would be complicated enough to scare out all other reviewers :)

vsemozhetbyt commented on test/common/README.md:
```diff
@@ -261,6 +261,124 @@ Platform normalizes the `pwd` command.
 
 Synchronous version of `spawnPwd`.
 
```
OK.

vsemozhetbyt replied:
However, this will break TOC structure and API reference alphabet principle. So I would abstain.

refack replied:
Ok. Let's see what others think

vsemozhetbyt commented on test/common/README.md:
````diff
@@ -261,6 +261,124 @@ Platform normalizes the `pwd` command.
+assert.strictEqual(unTagged, tagged);
+```
+
+### tagUnwrap()
````
It seems pretty usual term in text editors (also used in our style guide) for visual formatting. Your variants seem to suppose more aggressive editing.

vsemozhetbyt commented on test/common/README.md:
````diff
@@ -261,6 +261,124 @@ Platform normalizes the `pwd` command.
+assert.strictEqual(unTagged, tagged);
+```
+
+### tagLF()
````
This will break the parallel with tagCRLF (no keeping there).

refack commented:
Read your comments 👍 on all but https://github.com/nodejs/node/pull/13016#discussion_r116381744, think again, but it's your call.

vsemozhetbyt commented:
I've added a bit more explanation in doc (including intro note) and addressed some requests.

CI: https://ci.nodejs.org/job/node-test-pull-request/8070/

vsemozhetbyt commented on test/parallel/test-stdout-close-catch.js:
```diff
@@ -6,10 +6,12 @@ const child_process = require('child_process');
-            JSON.stringify(testScript) + ' | ' +
-            JSON.stringify(process.execPath) + ' ' +
-            '-pe "process.stdin.on(\'data\' , () => process.exit(1))"';
+const cmd = common.tagUnwrap`
```
Or with `util.inspect()` :)

vsemozhetbyt commented:
I've replaced `tagLF` and `tagCRLF` by the `tagLFy` and `tagCRLFy` to make them more verbal.

CI: https://ci.nodejs.org/job/node-test-pull-request/8095/

vsemozhetbyt commented:
@refack I mean https://en.oxforddictionaries.com/definition/-fy Does it make sense?

refack commented:
> @refack I mean https://en.oxforddictionaries.com/definition/-fy Does it make sense?

I'm just making faces :bowtie:
If the new names make you more comfortable, I approve.

vsemozhetbyt commented:
cc @nodejs/collaborators : this PR extends test `common` library and adds 4 new functions to many places of many tests. Does anybody object to landing? If no, I shall land tomorrow.

mcollina commented:
I am 👍  with the change, but 👎  on the amount of breakage that might introduce. I already had to backport a PR from master to v7 and soon v6, because of changes in the test harness.
I think we should have a solid plan to backport this to v6 _before_ we land this in master.

lpinca commented:
We keep adding more and more stuff to `common.js` and I honestly don't like this very much. It increases the cognitive load required to write tests. Tests should ideally only use an assertion library.

That said I have no objections on this PR.

vsemozhetbyt commented:
@mcollina  It is a pity how the backporting burden prevents us from lightening other burdens, but it seems we should respect this burden as we respect gravitation.

@lpinca I understand and I recall the recent attempts to slim down the `common`. But the help from these functions seems to surpass its cognitive load for me. However, this is too personal (I have been using some of them for a while), so I do not insist.

It seems we have sound (yet soft) objections against the both commits: the fist makes `common` more bloated, the second brings too much churn. So if there is no strong "pro" I shall abstain to land.

mcollina commented:
I'm in favor of this change, but only if it is backported asap to v6. Otherwise it creates too much churn.

vsemozhetbyt commented:
@mcollina I can do a backport PR for v6 after landing if this is not backported cleanly. How much asap this should be made?

mcollina commented:
cc @nodejs/lts for that answer.

refack commented:
> It increases the cognitive load required to write tests.

@lpinca I agree that it is very important to lower the barrier (cognitive or other) for writing test! But IMHO in this specific case the trade-off of cognitive load while _writing_ code is cognitive load while _reading/fixing_ code. There are good examples that definatly make the code much more readable, like [here](https://github.com/nodejs/node/pull/13016/files#diff-97207ba21d5e71cac474e66e4d5f4b64R57), and [here](https://github.com/nodejs/node/pull/13016/files#diff-9791fd6e3a86873c13ceb536ae734869)

@vsemozhetbyt will these tags benefit `/benchmarks/`?

vsemozhetbyt commented:
@refack 
>  will these tags benefit `/benchmarks/`?
I have not checked yet, I just want to see if this can be accepted for tests. But I planned to try to apply for benchmarks as well.

vsemozhetbyt commented:
As for tradeoffs... For writing, the load is eased by optionality of using this functions. For reading, the load is increased by the "what the ... this `tag*()` means" surprise effects :)

lpinca commented:
@refack I'm a bit unconvinced by your second example.

```js
{
  data: tagCRLFy`
          POST / HTTP/1.0
          Connection: keep-alive


  `
}
```

is actually harder for me to grok than

```js
{
  data: 'POST / HTTP/1.0\r\n' +
        'Connection: keep-alive\r\n' +
        '\r\n'
}
```

Anyway, to reiterate, I have nothing against these changes. I'm fine if this gets merged.

vsemozhetbyt commented:
@lpinca @refack Yes, the examples with some trailing line breaks look a bit awkward with this approach, but this is the price of the function universality. I can revert these changes if there is strong -1 here.

refack commented:
> @refack I'm a bit unconvinced by your second example.

fair enough...

refack commented:
> the examples with some trailing line breaks look a bit awkward with this approach, but this is the price of the function universality

To follow up on https://github.com/nodejs/node/pull/13016#discussion_r116380358 IMHO since HTTP will probably be the main use case for this one, we could add (in following PR) a #{HTTP_SECTION_END} symbol.

vsemozhetbyt commented:
@evanlucas I hear you.

evanlucas commented:
@vsemozhetbyt :] I definitely appreciate the effort put in

sam-github commented:
A couple thoughts:

1. I think the verbose cut-n-paste in the tests decreases their readability and maintainability (like cut-n-past in any other code base), and am generally in favour of test utility functions.

2. These specific changes appear to make the code longer/more verbose, and less clear in its intent, so I'm not particularly convinced they are an improvement.

3. For backporting, @nodejs/lts attempts to make sure that changes like this are backported so changes after it will cherry-pick clean. I can't speak for the WG, but I'm pretty sure we'd land this on 6.x if it lands on 7.x.

vsemozhetbyt commented:
Well, it seems this is a too controversial proposition. So I better close for now to not distract more collaborators anymore. Thank you for all the feedback and time. I hope this was not completely in vain)

refack commented:
IMHO the most controversial was `tagGlue`, but `tagCRLFy` (that should be named `tagHTTPfy`) and `tagLFy` have some very good uses. So maybe don't try to cover everything and focus on the really good ones:

1. https://github.com/nodejs/node/pull/13016/files#diff-97207ba21d5e71cac474e66e4d5f4b64
2. https://github.com/nodejs/node/pull/13016/files#diff-d4784c96a833a5b56576449c93b866fc
3. https://github.com/vsemozhetbyt/node/blob/306f2287ca5f40e85430ea988b3f74f1e55160d7/test/parallel/test-https-client-resume.js
4. https://github.com/vsemozhetbyt/node/blob/306f2287ca5f40e85430ea988b3f74f1e55160d7/test/parallel/test-https-eof-for-eom.js
5. https://github.com/vsemozhetbyt/node/blob/306f2287ca5f40e85430ea988b3f74f1e55160d7/test/parallel/test-repl-persistent-history.js
6. https://github.com/vsemozhetbyt/node/blob/306f2287ca5f40e85430ea988b3f74f1e55160d7/test/parallel/test-tls-over-http-tunnel.js
7. https://github.com/vsemozhetbyt/node/blob/306f2287ca5f40e85430ea988b3f74f1e55160d7/test/parallel/test-util-inspect.js

vsemozhetbyt commented:
@refack I shall ponder on this.
